### PR TITLE
[InstSimplify] Make sure the simplified value doesn't generate poison in threadBinOpOverSelect

### LINKED ISSUE #87042: [RISCV] Miscompile with -O3
Reduced test case:
```
#include <stdio.h>
#include <stdint.h>
int16_t g;
int32_t f;
uint32_t h;
uint64_t j = 6;
static uint64_t o(int16_t x, uint32_t *a, int16_t ac) {
  if (!x) j |= f;
  f = ac ? 7 % ac : 7;
  return j;
}
int main() {
  uint64_t v1 =  o(g, &h, 8);
  uint64_t v2 = o(v1, &h, 0);
  o(v2 >> 8, &h, 0);
  printf("%lu\n", j);
  return 0;
}
```

```
$ gcc -O0 -fsanitize=address,undefined test.c && ./a.out
7
$ bin/clang -O3 --target=riscv64-linux-gnu test.c
$ qemu-riscv64 -L /usr/riscv64-linux-gnu ./a.out
13
```

llvm version: ba6b2d22af177a72b132cdb8e9350a708f282d2c
qemu version: v8.2.0

### PR DESCRIPTION
**Pull request by dtcxzyw**
Alive2: https://alive2.llvm.org/ce/z/y_Jmdn
Fix https://github.com/llvm/llvm-project/issues/87042.

### CODE REVIEW

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 1]
```diff
@@ -440,7 +440,8 @@ static Value *threadBinOpOverSelect(Instruction::BinaryOps Opcode, Value *LHS,
     // Check that the simplified value has the form "X op Y" where "op" is the
     // same as the original operation.
     Instruction *Simplified = dyn_cast<Instruction>(FV ? FV : TV);
-    if (Simplified && Simplified->getOpcode() == unsigned(Opcode)) {
+    if (Simplified && Simplified->getOpcode() == unsigned(Opcode) &&
+        !Simplified->hasPoisonGeneratingFlags()) {
```
Comment (goldsteinn) [inline llvm/lib/Analysis/InstructionSimplify.cpp:444]: Shouldn't `isSafeToSpeculativelyExecute` suffice here?

Comment (dtcxzyw) [reply llvm/lib/Analysis/InstructionSimplify.cpp:444]: `isSafeToSpeculativelyExecute` only cares about immediate UB instead of poison.

Comment (goldsteinn) [reply llvm/lib/Analysis/InstructionSimplify.cpp:444]: why is the fix to prohibit any poison flag and not to just drop the poison generating flag?

Comment (dtcxzyw) [reply llvm/lib/Analysis/InstructionSimplify.cpp:444]: IIRC modifying an instruction is not allowed in InstSimplify :(

Comment (goldsteinn) [reply llvm/lib/Analysis/InstructionSimplify.cpp:444]: oh yeah xD

Maybe just move this to InstCombine? Or put equivelent code that drops poison generating flags there?

**Comment by goldsteinn** [comment]
LGTM

**Comment by nikic** [comment]
Can you please check what the impact of just deleting this special case entirely is? IIRC the last time I looked at this highly dubious code it didn't have any test coverage at all (though I might have added some at the time).

**Comment by dtcxzyw** [comment]
Emm, it would be better to move the logic into InstCombine.

**Comment by goldsteinn** [comment]
Is this dead?

**Comment by dtcxzyw** [comment]
> Is this dead?

No. TBH I prefer this patch because I think the regression is acceptable.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**